File: src/main/java/com/github/svegon/capi/event/render/BlockEntityRenderCallback.java
package com.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.client.render.VertexConsumerProvider;
import net.minecraft.client.util.math.MatrixStack;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

public interface BlockEntityRenderCallback {
    Event<BlockEntityRenderCallback> EVENT = EventFactory.createArrayBacked(BlockEntityRenderCallback.class,
            (blockEntity, tickDelta, matrices, vertexConsumers, callback) -> {},
            listeners -> (blockEntity, tickDelta, matrices, vertexConsumers, callback) -> {
        for (BlockEntityRenderCallback listener : listeners) {
            listener.onBlockEntityRender(blockEntity, tickDelta, matrices, vertexConsumers, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
    });

    void onBlockEntityRender(BlockEntity blockEntity, float tickDelta, MatrixStack matrices,
                             VertexConsumerProvider vertexConsumers, CallbackInfo callback);
}

File: src/client/kotlin/io/github/svegon/mclientapi/client/event/render/BlockEntityRenderCallback.kt
package io.github.svegon.mclientapi.client.event.render

import io.github.svegon.mclientapi.client.mixinterface.IRenderTickCounter
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.block.entity.BlockEntity
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.util.math.MatrixStack
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo

interface BlockEntityRenderCallback {
    fun onBlockEntityRender(
        blockEntity: BlockEntity, tickDelta: Float, matrices: MatrixStack,
        vertexConsumers: VertexConsumerProvider, callback: CallbackInfo
    )

    companion object {
        @JvmField
        val EVENT: Event<BlockEntityRenderCallback> = EventFactory.createArrayBacked(
            BlockEntityRenderCallback::class.java,
            object : BlockEntityRenderCallback {
                override fun onBlockEntityRender(
                    blockEntity: BlockEntity,
                    tickDelta: Float,
                    matrices: MatrixStack,
                    vertexConsumers: VertexConsumerProvider,
                    callback: CallbackInfo
                ) {}
            }
        ) { listeners: Array<BlockEntityRenderCallback> ->
            object : BlockEntityRenderCallback {
                override fun onBlockEntityRender(
                    blockEntity: BlockEntity,
                    tickDelta: Float,
                    matrices: MatrixStack,
                    vertexConsumers: VertexConsumerProvider,
                    callback: CallbackInfo
                ) {
                    for (listener in listeners) {
                        listener.onBlockEntityRender(blockEntity, tickDelta, matrices, vertexConsumers, callback)

                        if (callback.isCancelled) {
                            return
                        }
                    }
                }
            }
        }
    }
}
